### PR TITLE
#4678: Converting strings to Date when schema.type is Date within agg…

### PR DESCRIPTION
**Pull request by cjbland**
I added a check to see if the schema field is a `Date`. If so, convert the string into a `Date` object which the `mongodb` driver will accept for the aggregate function (it will not accept strings). I also added checks for the `createdAt` and `updatedAt` attributes and converted them to the actual column names.

### CODE REVIEW

**Comment by cjbland** [comment]
I need to keep working on this, I had forgotten about the other operators such as `$gt`, `$gte`, etc.

**Comment by flovilmart** [comment]
@cjbland no worry! Thanks for your contributions!

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 1]
```diff
@@ -565,34 +565,7 @@ export class MongoStorageAdapter implements StorageAdapter {
...
-            stage.$match = transformMatch;
-          }
-          else if (field === 'updatedAt') {
-            const transformMatch = Object.assign({}, stage.$match);
-            transformMatch._updated_at = stage.$match[field];
-            delete transformMatch.updatedAt;
-            stage.$match = transformMatch;
-          }
-        }
+        this._parseAggregateArgs(schema, stage.$match);
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Returned value seem unused.

Comment (cjbland) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Thanks for picking up on that...I was trying to hurry and pushed the repo by accident (muscle memory from my day job, haha). I just pushed the fix.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 2]
```diff
@@ -608,6 +597,55 @@ export class MongoStorageAdapter implements StorageAdapter {
       .catch(err => this.handleError(err));
   }
 
+  _parseAggregateArgs(schema: any, pipeline: any): any {
+    var rtnval = {};
+    if (Array.isArray(pipeline)) {
+      rtnval = [];
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: can we transform in 

```
return pipeline.map((value) => this._parseAggregateArgs(schema, value));
```

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 3]
```diff
@@ -608,6 +597,55 @@ export class MongoStorageAdapter implements StorageAdapter {
       .catch(err => this.handleError(err));
   }
 
+  _parseAggregateArgs(schema: any, pipeline: any): any {
+    var rtnval = {};
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: use const, and actually remove :)

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 4]
```diff
@@ -608,6 +597,55 @@ export class MongoStorageAdapter implements StorageAdapter {
...
+    if (Array.isArray(pipeline)) {
+      rtnval = [];
+      for (const field in pipeline) {
+        rtnval.push(this._parseAggregateArgs(schema, pipeline[field]));
+      }
+    }
+    else if (typeof pipeline === 'object') {
+      for (const field in pipeline) {
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: declare `const returnValue = {}` here and update all fields please

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 5]
```diff
@@ -608,6 +597,55 @@ export class MongoStorageAdapter implements StorageAdapter {
...
+        if (schema.fields[field] && schema.fields[field].type === 'Pointer') {
+          rtnval[`_p_${field}`] = `${schema.fields[field].targetClass}$${pipeline[field]}`;
+        }
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: throughout the codebase we always write

``` 
if  {

} else if ... {

} else if ... {

}
```

Can we keep the style please?

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 6]
```diff
@@ -608,6 +597,55 @@ export class MongoStorageAdapter implements StorageAdapter {
...
+    return rtnval;
+  }
+
+  _convertToDate(value: any): any {
+    if (typeof value === 'string') {
+      return new Date(value);
+    }
+
+    var rtnval = {}
+    for (const field in value) {
```
Comment (flovilmart) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: can we document what we expect the value to look like, I have trouble understanding what it does, and why we need the iterator, and recursion.

**Comment by flovilmart** [comment]
@cjbland also it seems that many tests are failing :) You can run them locally with `npm test spec/ParseQuery.Aggregate.spec.js` so only the aggregate suite is run!

**Comment by cjbland** [comment]
Thanks for the tip. I didn't realize the tests were failing but I see them now. I'll also fix the styling and take care of the other comments you've made. Thanks for taking your time to review this and sorry I'm not making it a little easier for you.

**Comment by flovilmart** [comment]
@dplewis do you want to take-over with @cjbland on this PR? Perhaps @cjbland , you could get back on the aggregate-recursive branch and see if your use case is solved. What were you planing @dplewis ? I know you've been busy lately with other things.

**Comment by dplewis** [comment]
@flovilmart I noticed 3-4 open issues and a lot of users requested transforming keys like pointers, default fields, and dates. I was trying to handle all use cases but I kinda lack complex aggregate queries to ensure it works.

The current transform code is getting to be too much to handle considering there are ~40 mongo aggregate stages and could use some refactoring.

The stages could be broken down as follows.
```
// { limit: 1 }
{ stageName: value}

// value can be any type or $field
// { group : { objectId: '' }
// { group: { objectId: '$name' }
{ stageName: { field: value }

// { match: { pointer: { $exist: true } } }
{ stageName: { field: { op: value } }

// group: { total: { $sum: '$score' } },
{ stageName: { newField: { op: value } }

// match: { $or: [{ score: { $gt: 15, $lt: 25 } }, { views: { $gt: 750, $lt: 850 } }]}
{ stageName: { op: value } 
```

There maybe more that I don't know of. I can submit a PR and worry about complex queries as they come along.

Here's an example of the most complex aggregate I've used. https://github.com/parse-community/parse-server/issues/1908#issuecomment-371589760

**Comment by cjbland** [comment]
I'm definitely open to collaborating. The recursive aggregate function I wrote seems to work, my issue is the test case... I don't have any experience writing these tests, but I feel like the previous tests' `TestObject`s are being returned when I wasn't expecting them. Any pointers on how best to debug something like that?

**Comment by dplewis** [comment]
@cjbland For sure, checkout the Contributing.MD for running test cases.
https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md

There is a function `loadTestData()`  that gets called before each test.

https://github.com/parse-community/parse-server/blob/master/spec/ParseQuery.Aggregate.spec.js#L20

Thats where your `TestObject`s are coming from. Keep in mind that those objects exists

**Comment by cjbland** [comment]
I added some comments to hopefully document my approach here. Personally, I'm not a fan of recursion, but this seems like the perfect candidate for it. The goal was to treat the query like a tree and traverse the tree to get to the leaf nodes. Once as the leaf node determine whether or not the value needed to be converted (based on the schema). I understand the structure isn't _exactly_ a tree, but that's the best way for me to describe my intent. This should also handle any of the aggregation cases (since we aren't looking specifically pipeline operators or comparison operators).

**Comment by dplewis** [comment]
@cjbland I like your approach better. Can you copy the tests from my branch to see if they run?

**Comment by cjbland** [comment]
Absolutely. Give me a few days, I have a lot going on this week.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js (hunk 6 above)
Comment (cjbland) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: Sorry for the newbie question...how do we take care of these? I've updated my branch to include your suggestions. Thanks.

**Comment by flovilmart** [comment]
@dplewis where do we stand on this one? I know you had some ideas around this.

**Comment by kevflynn** [comment]
is anybody going to fix this or should i assume no...?

**Comment by flovilmart** [comment]
@kevflynn we are actively discussing it, not sure where you’re coming from with that attitude. I you feel the resolution is too slow for you, feel free to propose another PR.

**Comment by kevflynn** [comment]
Wow, @flovilmart not sure why you assume that was ‘attitude’. Was a pretty fair question considering no activity in 21 days, right? Was purely intended to be informational/see if this PR was being abandon. Yikes.

**Comment by dplewis** [comment]
@flovilmart I think this is good for the 3 main code paths. Maybe a few nits

**Comment by flovilmart** [comment]
> no activity in 21 days, right

Besides a comment the day before, checking if the reviewer had anything else to add. But yes you’re right.

@dplewis,can you jump on a final review please?

**Comment by kevflynn** [comment]
fair enough. Thanks for the help with this

File: spec/ParseQuery.Aggregate.spec.js [hunk 7]
```diff
@@ -164,6 +164,32 @@ describe('Parse.Query Aggregate testing', () => {
...
+      const query = new Parse.Query(TestObject);
+      return query.aggregate(pipeline);
+    }).then((results) => {
+      const createdAt = new Date(obj1.createdAt);
+      expect(results[0].objectId.day).toEqual(createdAt.getUTCDate());
+      expect(results[0].objectId.month).toEqual(createdAt.getMonth() + 1);
+      expect(results[0].objectId.year).toEqual(createdAt.getUTCFullYear());
+      done();
+    }).catch((error) => {
+      console.log(error);
```
Comment (dplewis) [inline spec/ParseQuery.Aggregate.spec.js]: I see in the travis logs that this catch gets called.

**Comment by dplewis** [comment]
There is a test at should pass but fails. Transform keys works great but you have to transform values as well for fieldnames that start with $

**Comment by cjbland** [comment]
I'm looking at this tonight, sorry for taking so long we had a busy weekend.

**Comment by flovilmart** [comment]
@cjbland there is no pressure :) take the time it needs

File: spec/ParseQuery.Aggregate.spec.js (hunk 7 above)
Comment (dplewis) [reply spec/ParseQuery.Aggregate.spec.js]: @cjbland Can you remove the catch from this test? It does pass now

**Comment by cjbland** [comment]
I made the change that @dplewis suggested and added more tests for better code coverage. Hopefully this gets us a little closer.

**Comment by cjbland** [comment]
Is there documentation on which of these aggregate functions are supported by PostgreSQL? That seems to be why the latest tests failed for test 6417.3 - it didn't know how to handle the `$exists` operator.

**Comment by dplewis** [comment]
@cjbland that’s the reason why that test was excluded. Postgres only supports basic aggregates aka 3 stages no operators

**Comment by cjbland** [comment]
I'm happier with this now. Let me know what the next steps are. Thanks guys!

**Comment by dplewis** [comment]
@flovilmart I did some linter clean up can you do a final review?

**Comment by dplewis** [comment]
@cjbland Thanks for the PR

**Comment by cjbland** [comment]
No problem, thanks for your patience. I’d like to contribute more if needed.

**Comment by flovilmart** [comment]
@cjbland, thanks! Contributions can take many forms, and we always encourage people who want to start contributing a bit more to respond on some issues. If you feel you can address a few of the open ones, i’ll Gladly walk you through it and give the appropriate accesses so you can close / label them!

**Comment by EricNetsch** [comment]
@dplewis @flovilmart 
Looks like this was released into 2.8.3 but not seeing the new code on my side when updating/reinstalling.

Any ideas on how what's going on?
Thanks!

**Comment by flovilmart** [comment]
I wasn't released part of 2.8.3 as this version is a hotfix for compatibility with the JS SDK.

https://github.com/parse-community/parse-server/compare/2.8.2...2.8.3

**Comment by EricNetsch** [comment]
@flovilmart Got it. Saw it in the release commits so assumed it was part of 2.8.3 . 
Any idea on when we can release this fix? 

Thanks for clarifying,
screenshot 2018-08-11 17 47 47

**Comment by flovilmart** [comment]
We’re working on the 3.0 at the moment, and it’s likely gonna be the next release. Perhaps there will be a 2.8.4 before we cut the 3.0

**Comment by EricNetsch** [comment]
Got it thanks. 
Ahh I see I was looking at master commits since 2.8.3 lol. Confusing UX on github releases.
Appreciate it.

**Comment by flovilmart** [comment]
It’s always unlikely a release is ahead of master :)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0-beta.1](https://github.com/parse-community/parse-server/releases/tag/5.0.0-beta.1)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [5.0.0](https://github.com/parse-community/parse-server/releases/tag/5.0.0)

**Pull request merged**